analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -93,11 +54,30 @@ protected SonarAnalysisContextBase(SonarAnalysisContext analysisContext, TContex
         Context = context;
     }
 
-    public override bool TryGetValue<TValue>(SourceText text, SourceTextValueProvider<TValue> valueProvider, out TValue value) =>
-        AnalysisContext.TryGetValue(text, valueProvider, out value);
+    /// <param name="tree">Tree to decide on. Can be null for Symbol-based and Compilation-based scenarios. And we want to analyze those too.</param>
+    /// <param name="generatedCodeRecognizer">When set, generated trees are analyzed only when langauge-specific 'analyzeGeneratedCode' configuration property is also set.</param>
+    public bool ShouldAnalyzeTree(SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer) =>
+        (generatedCodeRecognizer is null || ShouldAnalyzeGenerated() || !tree.IsGenerated(generatedCodeRecognizer, Compilation))
+        && (tree is null || !IsUnchanged(tree));
 
-    public ProjectConfigReader ProjectConfiguration() =>
-        ProjectConfiguration(Options);
+    /// <summary>
+    /// Reads configuration from SonarProjectConfig.xml file and caches the result for scope of this analysis.
+    /// </summary>
+    public ProjectConfigReader ProjectConfiguration()
+    {
+        if (Options.SonarProjectConfig() is { } sonarProjectConfig)
+        {
+            return sonarProjectConfig.GetText() is { } sourceText
+                // TryGetValue catches all exceptions from SourceTextValueProvider and returns false when exception is thrown
+                && AnalysisContext.TryGetValue(sourceText, ProjectConfigProvider, out var cachedProjectConfigReader)
+                ? cachedProjectConfigReader
+                : throw new InvalidOperationException($"File '{Path.GetFileName(sonarProjectConfig.Path)}' has been added as an AdditionalFile but could not be read and parsed.");
+        }
+        else
+        {
+            return ProjectConfigReader.Empty;
+        }
+    }
 
     public bool IsScannerRun() =>
```

Comment:
This is now used only in this file and can be made private.

Reply:
Or removed as unneeded indirection :)